src/Microsoft.Health.Dicom.SqlServer/Features/Query/SqlQueryGenerator.cs

```diff
@@ -121,6 +132,85 @@ private void AppendFilterTable(string filterAlias)
             _stringBuilder.AppendLine($") {filterAlias}");
         }
 
+        private void AppendCustomTagTables()
+        {
+            Dictionary<CustomTagDataType, CustomTagLevel> queriedCustomTagDataTypesPerLevel = new Dictionary<CustomTagDataType, CustomTagLevel>();
+            foreach (CustomTagFilterDetails filterDetails in _queryExpression.QueriedCustomTagFilterDetails)
+            {
+                CustomTagDataType dataType;
+                CustomTagLimit.CustomTagVRAndDataTypeMapping.TryGetValue(filterDetails.Tag.GetDefaultVR().Code, out dataType);
+                if (!queriedCustomTagDataTypesPerLevel.TryAdd(dataType, filterDetails.Level) && queriedCustomTagDataTypesPerLevel[dataType] > filterDetails.Level)
+                {
+                    queriedCustomTagDataTypesPerLevel[dataType] = filterDetails.Level;
+                }
+            }
+
+            foreach (CustomTagFilterDetails filterDetails in _queryExpression.QueriedCustomTagFilterDetails)
+            {
+                CustomTagDataType dataType;
+                CustomTagLimit.CustomTagVRAndDataTypeMapping.TryGetValue(filterDetails.Tag.GetDefaultVR().Code, out dataType);
+                string customTagTableAlias = null;
+                _stringBuilder.Append("INNER JOIN ");
+                switch (dataType)
+                {
+                    case CustomTagDataType.StringData:
```

Comment:
Does this work for all level queries.

Reply:
Yes! I've added in tests to validate it as well.

---
In reply to: [582212608](https://github.com/microsoft/dicom-server/pull/567#discussion_r582212608) [](ancestors = 582212608)